File: examples/seo-site/app/products/[slug]/actions.go
// Package products_slug implements actions declared by the product route.
package products_slug

import (
	"errors"

	gb "github.com/Origens-Dev/gobeyond"
	contract "github.com/Origens-Dev/gobeyond/examples/seo-site/internal/gobeyondgen/contracts/actions/r_products_slug_3e2e8eb9_add_to_cart"
)

func AddToCart(_ *gb.ActionContext, input contract.Input) (contract.Output, error) {
	if input.ProductSlug == "" || input.Quantity < 1 || input.Quantity > 20 {
		return contract.Output{}, errors.New("productSlug and a quantity from 1 to 20 are required")
	}
	return contract.Output{Saved: true, CartItemCount: input.Quantity}, nil
}

File: examples/seo-site/app/products/[slug]/actions.ts
import {
  defineAction,
  schema,
  type InferActionInput,
  type InferActionOutput,
} from "@go-beyond/schema";

export const addToCart = defineAction({
  input: schema.object({
    productSlug: schema.string(),
    quantity: schema.integer(),
  }),
  output: schema.object({
    saved: schema.boolean(),
    cartItemCount: schema.integer(),
  }),
});

export type AddToCartInput = InferActionInput<typeof addToCart>;
export type AddToCartOutput = InferActionOutput<typeof addToCart>;
